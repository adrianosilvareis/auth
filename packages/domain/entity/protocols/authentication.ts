import { Accounts } from "./account"
import { AuthGroup } from "./auth-group"
import { uuid } from "./utils"

export enum AuthenticationStatusEnum {
  Online,
  Offline,
  Blocked,
  Created
}

export interface Authentication {
  id: uuid
  account: Accounts
  password: string
  attempts: number
  sessionLimit: number
  status: AuthenticationStatusEnum
  active: boolean
  authGroup: AuthGroup[]
  createdAt: Date
  updatedAt: Date
}

export type AuthenticationProperties = Omit<Authentication, "id">

export interface AuthenticationList {
  getList(): Promise<Authentication[]>
}

export interface AuthenticationByAccount {
  getByAccountId(accountId: uuid): Promise<Authentication>
}

export interface UpdateAuthentication {
  updateById(id: uuid, auth: AuthenticationProperties): Promise<Authentication>
}

export interface CreateAuthentication {
  create(auth: AuthenticationProperties): Promise<Authentication>
}

export interface CancelAuthentication {
  cancel(authenticationId: uuid): Promise<void>
}
