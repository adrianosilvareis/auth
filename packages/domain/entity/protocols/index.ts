export * from "./account"
export * from "./activity-log"
export * from "./auth-group"
export * from "./authentication"
export * from "./session"
export * from "./utils"
