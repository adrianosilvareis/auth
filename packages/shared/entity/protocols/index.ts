export * from './Account'
export * from './ActivityLog'
export * from './AuthGroup'
export * from './authentication'
export * from './Session'
export * from './Utils'