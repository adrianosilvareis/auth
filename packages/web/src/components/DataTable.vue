<template>
  <div class="form-floating mx-3 mt-3">
    <input
      id="search"
      class="form-control"
      type="text"
      placeholder="Search"
      v-model="search"
    />
    <label for="search">Search</label>
  </div>
  <table class="table">
    <thead>
      <tr>
        <th scope="col" v-for="(header, index) in headers" :key="index">
          {{ header.title }}
        </th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="(item, index) in filterdItems" :key="index">
        <td v-for="(header, index) in headers" :key="index">
          {{ item[header.key] }}
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script lang="ts">
import { Vue, Options } from "vue-class-component";

interface IHeaders<T> {
  title: string;
  key: keyof T;
}

@Options({
  name: "DataTable",
  props: {
    items: { type: Array, required: true },
    headers: { type: Array, required: true },
  },
})
export default class DataTable<T> extends Vue {
  items!: Array<T>;
  headers!: IHeaders<T>[];

  search = "";

  get filterdItems(): T[] {
    return this.items.filter((item) =>
      this.headers.some((header) =>
        item[header.key]
          .toString()
          .toUpperCase()
          .includes(this.search.toLocaleUpperCase())
      )
    );
  }
}
</script>
