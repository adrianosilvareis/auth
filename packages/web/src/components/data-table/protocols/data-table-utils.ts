export type Headers<T> = {
  title: string;
  key: keyof T;
};

export type ItemModel = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [k: string]: any;
};

export type ColumnsConfig<T> = (keyof T | Headers<T>)[];
