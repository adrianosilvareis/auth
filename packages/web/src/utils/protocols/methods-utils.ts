export interface ValidityDate {
  validity (createdOn: Date, now: Date, duration: number): boolean
}

export interface CalcAge {
  age (createdOn: Date, birthday: Date): string
}