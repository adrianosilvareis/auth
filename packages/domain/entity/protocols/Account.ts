import { Authentication } from "./authentication"
import { uuid } from "./utils"

export enum AccountStatusEnum {
  WAITING_FOR_APPROVAL,
  APPROVED,
  REPROVED,
  BLOCKED
}

export interface Address {
  id: uuid
  street: string // confidential data
  suite: string // confidential data
  city: string // confidential data
  zipCode: string // confidential data
}

export interface Accounts {
  id: uuid
  name: string // confidential data
  username: string // confidential data
  email: string // confidential data
  phone: string // confidential data
  address?: Address // confidential data,
  authentication?: Authentication
  createdAt: Date
  active: boolean // only approved status
  status: AccountStatusEnum
}

// get | update | create | cancel
export type AccountsProperties = Omit<Accounts, "id">
export type AccountsUpdates = Partial<Accounts>

export interface ListAccountByAuthGroup {
  listAccountByAuthGroupId(authGroupId: uuid): Promise<Accounts[]>
}

export interface ListAllAccount {
  listAll(): Promise<Accounts[]>
}

export interface GetAccountById {
  getById(accountId: uuid): Promise<Accounts>
}

export interface CancelAccountById {
  cancel(accountId: uuid): Promise<Accounts>
}

export interface UpdateAccountById {
  update(accountId: uuid, data: AccountsUpdates): Promise<Accounts>
}
