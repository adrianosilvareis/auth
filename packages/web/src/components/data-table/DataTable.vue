<template>
  <div class="row">
    <div class="col">
      <div class="form-floating mx-3 mt-3">
        <InputSearch v-model:search="search" />
      </div>
    </div>
    <div class="col d-flex justify-content-end">
      <div>button</div>
    </div>
  </div>
  <table class="table">
    <thead>
      <tr>
        <th scope="col" v-for="(header, index) in headers" :key="index">
          {{ header.title }}
        </th>
        <th v-if="hasActions">actions</th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="(item, index) in filteredItems" :key="index">
        <td v-for="(header, index) in headers" :key="index">
          <span v-if="editable && index === 0">
            <router-link :to="`${currentRouter}/${item[header.key]}`">
              {{ item[header.key] }}
            </router-link>
          </span>
          <span v-else>
            {{ item[header.key] }}
          </span>
        </td>
        <td v-if="hasActions">
          <slot name="actions" :item="item">
            <BtnRemove v-if="removable" @click="remove(item)" />
          </slot>
        </td>
      </tr>
      <tr v-if="notItems">
        <td class="text-center text-nowrap text-muted">
          <span v-if="search">Nenhum item encontrado para "{{ search }}"!</span>
          <span v-else>Nenhum item encontrado!</span>
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script lang="ts">
import { Vue, Options } from "vue-class-component"
import { ItemModel, Headers, ColumnsConfig } from "./protocols/data-table-utils"
import InputSearch from "@/components/form/InputSearch.vue"
import BtnRemove from "@/components/btn/BtnRemove.vue"

@Options({
  name: "DataTable",
  components: {
    InputSearch,
    BtnRemove
  },
  props: {
    items: { type: Array, required: true },
    columns: { type: Array, required: true },
    editable: { type: Boolean },
    removable: { type: Boolean }
  },
  emits: ["remove-item"]
})
export default class DataTable<T extends ItemModel> extends Vue {
  items!: Array<T>
  columns!: ColumnsConfig<T>
  editable!: boolean
  removable!: boolean

  search = ""
  get headers(): Headers<T>[] {
    return this.columns.map((column): Headers<T> => {
      if (typeof column === "string") {
        return {
          title: column,
          key: column
        }
      }
      if (typeof column === "object") {
        return {
          title: column.title,
          key: column.key
        }
      }
    })
  }

  get notItems(): boolean {
    return !this.filteredItems.length
  }

  get hasActions(): boolean {
    return !!this.$slots.actions || this.removable
  }

  get filteredItems(): T[] {
    if (!this.search) return this.items
    return this.items.filter((item) =>
      this.headers.some((header) =>
        item[header.key]
          .toString()
          .toUpperCase()
          .includes(this.search.toLocaleUpperCase())
      )
    )
  }

  get currentRouter(): string {
    return this.$route.fullPath
  }

  remove(item: T): void {
    this.$emit("remove-item", item)
  }
}
</script>
