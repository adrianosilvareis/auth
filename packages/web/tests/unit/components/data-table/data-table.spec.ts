import { shallowMount } from "@vue/test-utils";
import DataTable from "@/components/data-table/DataTable.vue";
import { Headers } from "@/components/data-table/protocols/data-table-utils";

describe("DataTable.vue", () => {
  it("renders props.columns and items values when passed", () => {
    const { wrapper, items, columns } = makeWrapper();
    expect(wrapper.text()).toMatch(columns[0].title);
    expect(wrapper.text()).toMatch(columns[1].title);
    expect(wrapper.text()).toMatch(items[0]["id"]);
    expect(wrapper.text()).toMatch(items[0]["title"]);
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders items filtered when search not null", async () => {
    const { wrapper, items } = makeWrapper();
    const inputSearch = wrapper.find('input[type="text"]')
    await inputSearch.setValue('any_title_2')
    
    const filterItems = items.filter(item => item.title.includes('any_title_2'))
    expect(wrapper.vm.filteredItems).toEqual(filterItems)
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders items filtered when search not null", async () => {
    const { wrapper, items } = makeWrapper();
    const inputSearch = wrapper.find('input[type="text"]')
    await inputSearch.setValue('any_title_2')
    
    const filterItems = items.filter(item => item.title.includes('any_title_2'))
    expect(wrapper.vm.filteredItems).toEqual(filterItems)
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders message on items filtered not found any items", async () => {
    const { wrapper, items } = makeWrapper();
    const inputSearch = wrapper.find('input[type="text"]')
    await inputSearch.setValue('any_title_3')
    
    expect(wrapper.text()).toMatch("Nenhum item encontrado para \"any_title_3\"!");
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders 'Nenhum item encontrado!' on empty items", async () => {
    const emptyItems:ModelTest[] = []
    const { wrapper, items } = makeWrapper(emptyItems);
    
    expect(wrapper.vm.filteredItems).toEqual(emptyItems)
    expect(wrapper.text()).toMatch("Nenhum item encontrado!");
    expect(wrapper.element).toMatchSnapshot();
  });
});

function makeWrapper(items: ModelTest[] = makeItems()) {
  const columns = makeColumns();
  const wrapper = shallowMount(DataTable, {
    props: { columns, items },
  });

  return {
    columns,
    items,
    wrapper,
  };
}

type ModelTest = {
  id: string;
  title: string;
}

function makeColumns():Headers<ModelTest>[]{
  return [
    { title: "#", key: "id" },
    { title: "My Title", key: "title" },
  ];
}

function makeItems():ModelTest[] {
  return [
    { 
      id: 'any_id',
      title: 'any_title'
    },
    { 
      id: 'any_id_1',
      title: 'any_title_1'
    },
    { 
      id: 'any_id_2',
      title: 'any_title_2'
    }
  ]
}