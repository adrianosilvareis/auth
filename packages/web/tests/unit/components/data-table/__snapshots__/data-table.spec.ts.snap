// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable.vue renders 'Nenhum item encontrado!' on empty items 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div
        class="form-floating mx-3 mt-3"
      >
        <input-search-stub
          search=""
        />
      </div>
    </div>
    <div
      class="col d-flex justify-content-end"
    >
      <div>
        button
      </div>
    </div>
  </div>
  <table
    class="table"
  >
    <thead>
      <tr>
        
        <th
          scope="col"
        >
          #
        </th>
        <th
          scope="col"
        >
          My Title
        </th>
        
        <!--v-if-->
      </tr>
    </thead>
    <tbody>
      
      
      <tr>
        <td
          class="text-center text-nowrap text-muted"
        >
          <span>
            Nenhum item encontrado!
          </span>
        </td>
      </tr>
    </tbody>
  </table>
  
</div>
`;

exports[`DataTable.vue renders props.columns and items values when passed 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div
        class="form-floating mx-3 mt-3"
      >
        <input-search-stub
          search=""
        />
      </div>
    </div>
    <div
      class="col d-flex justify-content-end"
    >
      <div>
        button
      </div>
    </div>
  </div>
  <table
    class="table"
  >
    <thead>
      <tr>
        
        <th
          scope="col"
        >
          #
        </th>
        <th
          scope="col"
        >
          My Title
        </th>
        
        <!--v-if-->
      </tr>
    </thead>
    <tbody>
      
      <tr>
        
        <td>
          <span>
            any_id
          </span>
        </td>
        <td>
          <span>
            any_title
          </span>
        </td>
        
        <!--v-if-->
      </tr>
      <tr>
        
        <td>
          <span>
            any_id_1
          </span>
        </td>
        <td>
          <span>
            any_title_1
          </span>
        </td>
        
        <!--v-if-->
      </tr>
      <tr>
        
        <td>
          <span>
            any_id_2
          </span>
        </td>
        <td>
          <span>
            any_title_2
          </span>
        </td>
        
        <!--v-if-->
      </tr>
      
      <!--v-if-->
    </tbody>
  </table>
  
</div>
`;
